### ISSUE TITLE
HomoNN: mismatch in the shape of predictions and labels due to TrainerBase.evaluation() function?

### ISSUE THREAD

**Issue by sousanunes** [FederatedAI/FATE#4594]
Hi!

This is more a clarification than a bug. But it seems quite serious so I decided to start an issue. I am targeting the HomoNN topology, with a ResNet implemented using [this example notebook](https://github.com/FederatedAI/FATE/blob/master/doc/tutorial/pipeline/nn_tutorial/Resnet-example.ipynb). I run into the following error in the trainer_base.py file:

![image](https://user-images.githubusercontent.com/5367028/214922948-1e0df82e-88d2-4c0c-8d44-ad2950534ea2.png)

**Reproducing the error**

1) In a standalone_fate v1.10.0 docker image, open [said notebook](https://github.com/FederatedAI/FATE/blob/master/doc/tutorial/pipeline/nn_tutorial/Resnet-example.ipynb).

Pertinently, note that the ResNet implementation has the following **forward function**:

```
def forward(self, x):
    if self.training:
        return self.classifier(self.resnet(x))
    else:
        return self.softmax(self.classifier(self.resnet(x)))`
```

2) To run the notebook, I replaced the training data by a small set of images belonging to two classes (0/1). I followed the standard pytorch ImageFolder format and replaced the data paths appropriately in the notebook (the issue happens with any data, so use any small set you have at hand):

- here: `ds.load('../../../../examples/data/cifar10/train/')`
- and here: `data_path = fate_project_path + '/examples/data/cifar10/train'`

3) To adapt to the two-class dataset, change the ResNet class from `self.classifier = t.nn.Linear(1000, 10)` to `self.classifier = t.nn.Linear(1000, 2)`.

4) When I run `pipeline.fit()` as it is, everything runs fine and life is good (you can run it with 2 epochs to make it faster).

5) When I add the parameter `validation_freqs=1` as an input to `TrainerParam` in the `nn_component`:

```
model = t.nn.Sequential(
    t.nn.CustModel(module_name='resnet', class_name='Resnet')
)

nn_component = HomoNN(name='nn_0',
                      model=model, 
                      loss=t.nn.CrossEntropyLoss(),
                      optimizer = t.optim.Adam(lr=0.001, weight_decay=0.001),
                      dataset=DatasetParam(dataset_name='image'),  # 使用自定义的dataset
                      trainer=TrainerParam(
                          trainer_name='fedavg_trainer', 
                          epochs=2, 
                          batch_size=1024, 
                          data_loader_worker=8, 
                          validation_freqs=1),
                      torch_seed=100
                      )
```

and I then try to run  `pipeline.fit()` in the guest node, I get the above-mentioned error: the mismatch in the shape of predictions and labels inside TrainerBase.evaluation(). (Note: my tiny training set has 160 instances.)

**After some investigation**

When the `validation_freqs` parameter is not provided, **no validation is done**. So the problem doesn't arise!

When performing validation though, it seems that the `evaluation()` function in the [`TrainerBase`](https://github.com/FederatedAI/FATE/blob/master/python/federatedml/nn/homo/trainer/trainer_base.py) class expects the predictions of two-class models to have shape `(n_samples, 1)`. However, the predictions generated by the ResNet have shape `(n_sample, n_classes)`, in this case `(n_samples, 2)`. These predictions are the output of the softmax in the forward function, mentioned above in point 1.

I think the fix would be to go from this:

![image](https://user-images.githubusercontent.com/5367028/214925784-c6e2b45d-0623-4623-8197-093f2dd6d2c2.png)

to this:

![image](https://user-images.githubusercontent.com/5367028/215048293-59340b33-2caa-43f8-aba0-f61cd5c8bb20.png)

in the `TrainerBase.evaluation()` function. I.e. we would take the class 1 probability. However, I'm hesitant to do the pull request since I am new to this (awesome) code and don't understand the full implications of this change.

Could you please help me understand what is happening and if this make sense? Am I missing something really obvious?

Thanks and apologies for the length!

**Environment**
Using [standalone_fate v1.10.0](https://hub.docker.com/r/federatedai/standalone_fate) from docker hub.

**Comment by talkingwallace** [comment]
Hi! Thanks for your feedback, by default the fedavg trainer will automatically infer the task type, ie., binary or multi-classification, according to the shape of the predicted result.  The binary task evaluation expects a shape of (n, 1) tensor to run the evaluation.
 
You can set this parameter to ‘binary': 
https://github.com/FederatedAI/FATE/blob/master/python/federatedml/nn/homo/trainer/fedavg_trainer.py#L44
and let your network outputs (n, 1) tensors, or set this parameter to 'multi'.

Thanks for your feedback and PR. We will update the codes the improve the auto-infer. :D.

**Comment by sousanunes** [comment]
Hi Wallace!

Thanks for getting back to me!

The automatic task type inference work fine. It infers binary correctly.

The issue is exactly what you said: it expects a (n,1) tensor when class predictions in pytorch are supposed to have shape (n, n_classes) (E.g. [check out line 206 of pytorch's ResNet](https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py)). Why should we have to implement an exception for the binary case?

I already adapted my code to handle that, no problem! But it's likely that someone might have the same assumption as me in the future.

Thanks again :)

**Comment by talkingwallace** [comment]
Hi!
The reasons why the Homo-nn trainer expects (n, 1) tensors for binary tasks are that:
1) all of the FATE binary-classification algorithms (like tree algorithms, and linear models) output (n, 1) tensors as the predicted result. We need to align formats with other algorithms.
2) In many industrial binary-classification models implemented by PyTorch (like Deep-FM) uses Sigmoid() to yield the predicted prob, and this will finally output (n, 1) tensors as results.

I hope these can help with your confusion. :D
